File: dsa-problems/Bitwise/solutions/cpp/BIT-011-toggle-ranges-min-flips.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int toggleRangesMinFlips(vector<int>& A, vector<int>& B) {
        int n = A.size();
        int count = 0;
        int prevDiff = 0;
        
        for (int i = 0; i < n; i++) {
            int currDiff = A[i] ^ B[i];
            if (currDiff == 1 && prevDiff == 0) {
                count++;
            }
            prevDiff = currDiff;
        }
        return count;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    if (!(cin >> n)) return 0;
    
    vector<int> A(n);
    for (int i = 0; i < n; i++) cin >> A[i];
    vector<int> B(n);
    for (int i = 0; i < n; i++) cin >> B[i];

    Solution solution;
    cout << solution.toggleRangesMinFlips(A, B) << "\n";
    return 0;
}

File: dsa-problems/Bitwise/solutions/java/BIT-011-toggle-ranges-min-flips.java
import java.util.*;

class Solution {
    public int toggleRangesMinFlips(int[] A, int[] B) {
        int n = A.length;
        int count = 0;
        int prevDiff = 0;
        
        for (int i = 0; i < n; i++) {
            int currDiff = A[i] ^ B[i];
            
            // If current is mismatch (1) and previous was match (0),
            // we have entered a new island of mismatches.
            // We greedily "start" a flip here. The flip "continues"
            // until the mismatch run ends, handled implicitly.
            if (currDiff == 1 && prevDiff == 0) {
                count++;
            }
            prevDiff = currDiff;
        }
        
        return count;
    }
}

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextInt()) return;
        int n = sc.nextInt();
        int[] A = new int[n];
        for (int i = 0; i < n; i++) A[i] = sc.nextInt();
        int[] B = new int[n];
        for (int i = 0; i < n; i++) B[i] = sc.nextInt();

        Solution solution = new Solution();
        System.out.println(solution.toggleRangesMinFlips(A, B));
        sc.close();
    }
}
